src/kem/kyber/matrix.rs

```diff
@@ -0,0 +1,113 @@
+use super::{
+    arithmetic::{
+        add_to_ring_element, barrett_reduce, montgomery_reduce, to_standard_domain,
+        PolynomialRingElement,
+    },
+    ntt::{invert_ntt_montgomery, ntt_multiply},
+};
+
+#[inline(always)]
+pub(in crate::kem::kyber) fn compute_message<const K: usize>(
+    v: &PolynomialRingElement,
+    secret_as_ntt: &[PolynomialRingElement; K],
+    u_as_ntt: &[PolynomialRingElement; K],
+) -> PolynomialRingElement {
+    let mut result = PolynomialRingElement::ZERO;
+
+    for i in 0..K {
+        let product = ntt_multiply(&secret_as_ntt[i], &u_as_ntt[i]);
+        result = add_to_ring_element::<K>(result, &product);
+    }
+
+    result = invert_ntt_montgomery::<K>(result);
+
+    for i in 0..result.coefficients.len() {
```

Comment:
Let's use the constant size `COEFFICIENTS_IN_RING_ELEMENT` here to avoid the `len()`.